### PR TITLE
Show incompatible tasks in CC report

### PR DESCRIPTION
**Pull request by abstratt**
Issue: #21290
Related: gradle/configuration-cache-report#18

### Summary of changes

- adapted collection of incompatible tasks so reason is also provided, to be included in the report
- include incompatible tasks in the CC report
- changed tracking of incompatible tasks to be based on PropertyTrace (likely a PropertyTrace.Task variant) instead of task paths
- added test coverage to verify incompatible tasks appear in the generated report
- generalized checking of inputs so the verification can be shared between inputs and incompatible tasks

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheIncompatibleTasksIntegrationTest.groovy [hunk 1]
```diff
@@ -60,6 +61,30 @@ class ConfigurationCacheIncompatibleTasksIntegrationTest extends AbstractConfigu
         '''
     }
 
+    def "configuration cache report includes incompatible tasks"() {
+        given:
+        buildFile '''
+        task reportedlyIncompatible {
+            notCompatibleWithConfigurationCache("declaring myself as not compatible")
+        }
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheIncompatibleTasksIntegrationTest.groovy:69]: 💅  Indent

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/ConfigurationCacheProblems.kt [hunk 2]
```diff
@@ -145,17 +154,43 @@ class ConfigurationCacheProblems(
...
+    private
+    fun reportIncompatibleTask(path: String, reason: String) {
+        val problem = problemFactory
+            .problem {
+                text("task ")
+                reference(path)
+                text(" is incompatible with the configuration cache. ")
+                text("Reason: ")
+                text(reason)
+                text(".")
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/ConfigurationCacheProblems.kt:166]: 💅 A single interpolated string would do

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/ConfigurationCacheProblems.kt [hunk 3]
```diff
@@ -145,17 +154,43 @@ class ConfigurationCacheProblems(
...
     override fun onProblem(problem: PropertyProblem) {
         onProblem(problem, ProblemSeverity.Failure)
     }
 
     private
     fun onProblem(problem: PropertyProblem, severity: ProblemSeverity) {
         if (summarizer.onProblem(problem, severity)) {
-            report.onProblem(problem)
+            when (severity) {
+                ProblemSeverity.Warning -> report.onIncompatibleTask(problem)
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/ConfigurationCacheProblems.kt:183]: ❔ It feels like there should be a more specific mechanism to distinguish between incompatible task problems. Not all warnings are necessarily incompatible task things
Something like `DiagnosticKind` seems a better fit here.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/JsonModelWriter.kt [hunk 4]
```diff
@@ -190,6 +192,12 @@ class JsonModelWriter(val writer: Writer) {
                 property("type", trace.type.name)
             }
 
+            is PropertyTrace.TaskPath -> {
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/JsonModelWriter.kt:195]: ❔ Why couldn't we reuse the `PropertyTrace.Task` here?

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/ConfigurationCacheProblems.kt (hunk 3 above)
Comment (abstratt) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/ConfigurationCacheProblems.kt]: Good point. This change was reverted, at some point it seemed necessary, but with the current state of this PR that does not seem required any longer

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/JsonModelWriter.kt (hunk 4 above)
Comment (abstratt) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/JsonModelWriter.kt]: Because it would require a `Task` object, and one is not readily available in the context where we collect incompatible tasks. But is not too far away either. Will commit changes to use `Task` instead of `TaskPath`.

Comment (abstratt) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/JsonModelWriter.kt]: `TaskPath` is gone

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblems.kt [hunk 5]
```diff
@@ -99,7 +103,7 @@ class ConfigurationCacheProblems(
     var updatedProjects = 0
 
     private
-    var incompatibleTasks = newConcurrentHashSet<String>()
+    var incompatibleTasks = newConcurrentHashSet<PropertyTrace>()
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblems.kt:106]: 🤔 Can this be a `val` ?

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/configurationcache/ConfigurationCacheProblemsFixture.groovy [hunk 6]
```diff
@@ -465,6 +491,7 @@ ${text}
         return new ProblemsSummary(totalProblems, problems.size(), problems)
     }
 
+    @ToString(includeNames=true)
```
Comment (alllex) [inline testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/configurationcache/ConfigurationCacheProblemsFixture.groovy:494]: 👍

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt [hunk 7]
```diff
@@ -109,58 +110,77 @@ data class StructuredMessage(val fragments: List<Fragment>) {
...
+    }
 
-    object Gradle : PropertyTrace()
+    object Gradle : PropertyTrace() {
+        override fun toString(): String = asString()
+    }
 
-    class BuildLogic(
+    data class BuildLogic(
         val source: DisplayName,
```
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt:122]: ❔  My only concern here is expecting this very general interface as a field. Currently, it seems not to be a problem, but it might be possible in the future to accidentally pass something here that would not necessarily be designed for being a "data" object. I was looking at the `DisplayName` implementations and their abundance.

**Review by alllex — APPROVED** [review]
LGTM!

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt (hunk 7 above)
Comment (abstratt) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt:122]: Would providing factory methods or additional constructors that accepted `Location` and `UserCodeSource` (and then extracted display name and line number, then narrowing the kind of display names we accept here) alleviate your concern?

Comment (alllex) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt:122]: I would say yes 👍

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt [hunk 8]
```diff
@@ -109,58 +110,77 @@ data class StructuredMessage(val fragments: List<Fragment>) {
...
         val name: String,
         val trace: PropertyTrace
     ) : PropertyTrace() {
         override val containingUserCode: String
             get() = trace.containingUserCode
+
+        override fun toString(): String = asString()
     }
 
-    override fun toString(): String =
```
Comment (bamboo) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt:163]: 💅 Since the `data class` change requires all types to implement `toString` now, wouldn't it make sense to declare `toString` as abstract at this level?

File: platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/TaskNodeCodec.kt [hunk 9]
```diff
@@ -230,11 +230,12 @@ suspend fun <T> T.withTaskOf(
...
 ) where T : IsolateContext, T : MutableIsolateContext {
     withIsolate(IsolateOwners.OwnerTask(task), codec) {
-        withPropertyTrace(PropertyTrace.Task(taskType, task.identityPath.path)) {
+        val taskTrace = PropertyTrace.Task(taskType, task.identityPath.path)
+        withPropertyTrace(taskTrace) {
             if (task.isCompatibleWithConfigurationCache) {
                 action()
             } else {
-                forIncompatibleType(task.identityPath.path, action)
+                forIncompatibleTask(taskTrace, task.reasonTaskIsIncompatibleWithConfigurationCache.get(), action)
```
Comment (bamboo) [inline platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/TaskNodeCodec.kt:238]: 💅 The current `trace` can be accessed from the context via the `trace` property, no need to declare a variable to hold the taskTrace

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/configurationcache/ConfigurationCacheProblemsFixture.groovy [hunk 10]
```diff
@@ -173,8 +176,9 @@ final class ConfigurationCacheProblemsFixture {
         } else {
             assertNoProblemsSummary(result.output)
         }
-        // TODO:bamboo avoid reading jsModel twice when asserting on problems AND inputs
-        assertInputs(result.output, rootDir, spec.inputs)
+        // TODO:bamboo avoid reading jsModel twice when asserting on problems AND inputs AND incompatible tasks
```
Comment (bamboo) [inline testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/configurationcache/ConfigurationCacheProblemsFixture.groovy:179]:
```suggestion
        // TODO:bamboo avoid reading jsModel more than once when asserting on problems AND inputs AND incompatible tasks
```

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblems.kt [hunk 11]
```diff
@@ -154,6 +163,20 @@ class ConfigurationCacheProblems(
         }
     }
 
+    private
+    fun reportIncompatibleTask(trace: PropertyTrace, reason: String) {
+        val problem = problemFactory
+            .problem {
+                text(trace.containingUserCode)
+                text(" is incompatible with the configuration cache. Reason: $reason.")
```
[the remark this answers is not in the document]
Comment (abstratt) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblems.kt:171]: I figured we should avoid just concatenating sentences as we don't know the form of the user provided reason. 

Also, isn't `reference` meant for symbolic strings, whereas incompatibility reasons are user-facing messages in plain English?

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblems.kt [hunk 12]
```diff
@@ -154,6 +163,20 @@ class ConfigurationCacheProblems(
         }
     }
 
+    private
+    fun reportIncompatibleTask(trace: PropertyTrace, reason: String) {
+        val problem = problemFactory
+            .problem {
+                text(trace.containingUserCode)
```
[the remark this answers is not in the document]
Comment (abstratt) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblems.kt:170]: That results in nested quotes:

'task \`:declared\` of type \`Broken\`' is incompatible with the configuration cache. Reason: ...

Screenshot 2024-06-12 at 14 39 56

File: platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/TaskNodeCodec.kt (hunk 9 above)
Comment (abstratt) [reply platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/TaskNodeCodec.kt:238]: I wondered whether a reentrant call could set a different trace, but I can't really see now how that would happen

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/initialization/ConfigurationCacheProblemsListener.kt [hunk 13]
```diff
@@ -134,19 +134,28 @@ class DefaultConfigurationCacheProblemsListener internal constructor(
...
+                location
+            }
+
+            is PropertyTrace.Task -> {
+                location
+            }
+
+            else -> {
+                locationForTask(task)
+            }
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/initialization/ConfigurationCacheProblemsListener.kt:149]:
```suggestion
            is PropertyTrace.BuildLogic -> location
            is PropertyTrace.Task -> location
            else -> locationForTask(task)
```

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblems.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (bamboo) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblems.kt:171]: > isn't reference meant for symbolic strings

Yeah, perhaps it would be an abuse.

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt [hunk 14]
```diff
@@ -107,60 +110,101 @@ data class StructuredMessage(val fragments: List<Fragment>) {
...
+        override fun toString(): String = asString()
+        override fun equals(other: Any?): Boolean = other === this
+        override fun hashCode(): Int = 0
+    }
 
-    object Gradle : PropertyTrace()
+    object Gradle : PropertyTrace() {
+        override fun toString(): String = asString()
+        override fun equals(other: Any?): Boolean = other === this
+        override fun hashCode(): Int = 0
```
Comment (bamboo) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt:130]: 💭 It feels cleaner to use a different hash code here.
```suggestion
        override fun hashCode(): Int = 1
```

**Review by bamboo — CHANGES_REQUESTED** [review]

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt (hunk 14 above)
Comment (abstratt) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt:130]: @bamboo What about 

``` 
override fun hashCode(): Int = Gradle::class.hashCode()
```

?

Comment (bamboo) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt]: ❓ By the way, aren't the default implementations of `equals` and `hashCode` enough? Why do they have to be abstract in the supertype?

Comment (abstratt) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt]: They are enough, indeed, at least of the `object`-based implementations. I wanted to signal that subclasses need to override them (as we do for `toString()`), even if all current subclasses did already get proper implementations (via `object` or `data class`), but if any of them needed to stop being a `data class` or `object`, we could forget to implement `equals` and `hashCode`.

Comment (bamboo) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt]: > I wanted to signal that subclasses need to override them
Ok 👍
